File: c/0027_RemoveElement.c
// Problem: Remove Element (https://leetcode.com/problems/remove-element/)
// Difficulty: Easy
// Date: 2021-11-24

/*
Given an integer array nums and an integer val, remove all occurrences of val in
nums in-place. The relative order of the elements may be changed.

Since it is impossible to change the length of the array in some languages, you
must instead have the result be placed in the first part of the array nums. More
formally, if there are k elements after removing the duplicates, then the first
k elements of nums should hold the final result. It does not matter what you
leave beyond the first k elements.

Return k after placing the final result in the first k slots of nums.

Do not allocate extra space for another array. You must do this by modifying the
input array in-place with O(1) extra memory.
*/

int removeElement(int* nums, int numsSize, int val){
    int i = 0;
    numsSize--;
    while (i <= numsSize) {
        if (nums[i] == val) {
            nums[i] = nums[numsSize]; // move target to the end
            numsSize--; // "discard" the last value
            i != 0 && i--; // decrement i in case the swapped value is the same as target
            continue;
        }
        i++;
    }
    return numsSize + 1;
}

File: py/0027_RemoveElement.py
# 27. Remove Element

from typing import List


class Solution:
    def removeElement(self, nums: List[int], val: int) -> int:
        j = len(nums) - 1
        for i in reversed(range(len(nums))):
            if nums[i] == val:
                nums[i], nums[j] = nums[j], nums[i]
                j -= 1
        return j + 1


if __name__ == "__main__":
    puzzles = [
        ([3, 2, 2, 3], 3),
        ([0, 1, 2, 2, 3, 0, 4, 2], 2),
        ([], 1),
        ([0], 1),
        ([1], 1)
    ]
    for puzzle in puzzles:
        print(Solution().removeElement(*puzzle))
        print(puzzle)

"""
Runtime
35
ms
Beats
73.81%
Memory
16.63
MB
Beats
22.20%
"""
